lib: ring_buffer: add raw byte access mode

nordic-krch opened the pull request:
Wanted to use ring buffer but found out that existing ring buffer is not a generic one by tailored to work in data item mode (data + 32b metadata). To avoid confusion, I've extended API to provide 'raw bytes' mode. Hopefully, that is less confusing than yet-another-ring-buffer.

Extended ring buffer to allow storing raw bytes in it. API has been
extended keeping 'data item' mode untouched.

Additionally, added API which allows direct operation on ring buffer memory with following scheme:

Putting data to a ring buffer:
- allocate buffer from ring buffer - get number of granted bytes in return (sys_ring_buf_bytes_alloc).
- write data to that buffer (e.g. pass buffer to DMA)
- claim number of valid bytes written which is equal or less than allocated amount (sys_ring_buf_bytes_put). 

Getting data from a ring buffer:
- get buffer with available data (sys_ring_buf_bytes_get).
- process data
- free processed data (equal or less than returned by sys_ring_buf_bytes_free).

Alternatively API provides functions for copying bytes in and out (sys_ring_buf_bytes_cpy_put/get).

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
lib: ring_buffer: add raw byte access mode

Extended ring buffer to allow storing raw bytes in it. API has been
extended keeping 'data item' mode untouched.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pfalcon commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -21,27 +22,51 @@ by its memory address.
+retriever is not large enough to hold the data item's data, the dequeue
+operation fails.
+
+Raw mode
```
So, why "raw" instead of more descriptive "bytes"?

pfalcon commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -114,6 +140,15 @@ which can be accessed using less efficient modulo operations.
     SYS_RING_BUF_DECLARE_SIZE(my_ring_buf, MY_RING_BUF_WORDS);
 
+The following code defines a ring buffer with an arbitrary-sized data buffer,
+which can be accessed using less efficient modulo operations. Ring buffer is
```
Are you sure reference to "less efficient modulo operations" should be here?

pfalcon commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -144,6 +179,20 @@ can be specified.
+
+.. code-block:: c
+
+    u8_t my_data[MY_DATA_WORDS];
```
MY_DATA_WORDS ?

pfalcon commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -144,6 +179,20 @@ can be specified.
+    u8_t my_data[MY_DATA_WORDS];
+    size_t ret;
+
+    ret = sys_ring_buf_raw_put(&ring_buf, my_data, SIZE_OF(my_data));
```
What is SIZE_OF() ?

pfalcon commented on include/ring_buffer.h:
```diff
@@ -161,6 +187,11 @@ static inline int sys_ring_buf_space_get(struct ring_buf *buf)
  * being preempted or by using a mutex to govern writes to the ring buffer.
  *
+ * @warning
+ * Ring buffer instance should not mix raw data (@ref sys_ring_buf_raw_put and
```
Maybe this warning should be added to docs too?

pfalcon commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +79,50 @@ int sys_ring_buf_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+	cpy_size = size > space ? space : size;
+
+	for (i = 0; i < cpy_size; i++) {
+		index = (i + buf->tail) % buf->size;
```
Ok, so please don't use "expensive modulo operations", please use `if` to check for overflow instead.

pfalcon commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +79,50 @@ int sys_ring_buf_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+		index = (i + buf->tail) % buf->size;
+		buf->buf.buf8[index] = data[i];
+	}
+	buf->tail = (buf->tail + cpy_size) % buf->size;
```
Don't use %.

pfalcon commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +79,50 @@ int sys_ring_buf_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+	cpy_size = size > space ? space : size;
+
+	for (i = 0; i < cpy_size; ++i) {
+		index = (i + buf->head) % buf->size;
```
Don't use %.

pfalcon commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +79,50 @@ int sys_ring_buf_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+		index = (i + buf->head) % buf->size;
+		data[i] = buf->buf.buf8[index];
+	}
+	buf->head = (buf->head + cpy_size) % buf->size;
```
Don't use %.

pfalcon reviewed (dismissed):
> than yet-another-ring-buffer.

That's truly heroic deed (no irony intended), most people so far: a) wrote yet-another-ring-buffer; b) stored bytes as 32-bit words; c) went out of their way to pack bytes into 32-bit words.

It's especially nice to see docs updated at once - but there's a few of copy-paste errors, please fix.

And most importantly - please don't use "expensive modulo operations", there's no need to ;-).

nordic-krch commented:
@pfalcon thanks for review. I've applied you comments but in many cases i've changed the implementation (see updated PR description) so please re-review.

pfalcon commented:
@nordic-krch :

> e.g. pass buffer to DMA

That's quite a turn for this API! ;-)

Pre-comment from just reading description:

> allocate  (sys_ring_buf_bytes_alloc).

> claim  (sys_ring_buf_bytes_put).

I'm not sure that "sys_ring_buf_bytes_put" describes the purpose ("claim") well.


>    get buffer with available data (sys_ring_buf_bytes_get).
>    free processed data (equal or less than returned by sys_ring_buf_bytes_get).

Here name of func just duplicated.



That's pre-comments, will look in detail later. Thanks for the update.

pfalcon commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -21,27 +22,66 @@ by its memory address.
+operation fails.
+
+Byte mode
+========
```
"=" should be of the same width as the heading.

nordic-krch replied:
done

pfalcon commented on include/ring_buffer.h:
```diff
@@ -116,6 +149,45 @@ static inline void sys_ring_buf_init(struct ring_buf *buf, u32_t size,
+ *             Ring buffer
+ * @return 1 if empty, 0 otherwise.
+ */
+static inline int sys_ring_buf_custom_is_empty(u32_t head, u32_t tail)
```
I believe the latest convention for internal functions is "z_" prefix (which makes no sense, as that should be *public* prefix, but d'oh, it's Zephyr). @nashif, @andrewboie for consult.

nashif replied:
> I believe the latest convention for internal functions is "z_" prefix (which makes no sense, as that should be public prefix, but d'oh, it's Zephyr). @nashif, @andrewboie for consult.

@pfalcon this is not funny anymore. This was not made up or decided by me or by @andrewboie, this was a project wide decision. If you have problems with that, you need to bring it forward to the many forums we have, but having these random "unrelated" comments across PRs and Issues about your dislike for certain things is not really helping anyone.

pfalcon replied:
@nashif: It seems there's a problem with "accepting criticism". As for "project wide decision", I'm for one wasn't aware of it (like, I didn't see anything in the devel mailing list, or via other "development communication channels"). The only way I'm in loop on it is that in some of @andrewboie's reviews he told me that. That's why I called him and you to confirm if here "sys_ring_buf_custom_is_empty" should instead be "z_sys_ring_buf_is_empty". So, I'm puzzled with your reaction.

nashif replied:
> It seems there's a problem with "accepting criticism".

I have no problem with criticism, but "d'oh, it's Zephyr" is being sarcastic in a very bad way. As I said, if you have an issue with something, raise it in the right forums and not in random comments across the github project which leads to off-topic discussions and confusion.

There are multiple meetings per week where such issues are being discussed, and everyone is welcome to join them, they are all listed in the wiki. If you want every single decision to be sent to the mailing list, that is a proposal you can make, there is nothing that we are trying to hide, but there is a limit to the channels and methods something can be communicated.

Again, if you think the project is not communicating enough and that we should over-communicate, please raise this in the relevant forums.

pfalcon replied:
@nashif: 

> but "d'oh, it's Zephyr" is being sarcastic in a very bad way.

Oh, I'm sorry then, I didn't think your maintainership feelings would be hurt.

> As I said, if you have an issue with something

Let's see. My RFC on namespacing headers/APIs in Zephyr: https://lists.zephyrproject.org/g/devel/topic/16760352#343 , dated 2017-04-01 (not an April 1st joke!). Using "z_" was discussed as some opportunity (in https://lists.zephyrproject.org/g/devel/message/361 specifically). It generated some discussion, but little support. Fast forward a year-something, and turns out that literally the opposite to this RFC was decided: it's now possible to tell if function belongs to Zephyr ... if users are not supposed to use such a function! My comment might be not funny, but this situation is definitely is. And my comment was intended to be just the same - ironic at best, not sarcastic.

> If you want every single decision to be sent to the mailing list

Well, currently I'm just trying to help with putting decisions into practice, which requires help from decision-makers, but so far we discuss procedural matters instead.

pfalcon commented on include/ring_buffer.h:
```diff
@@ -116,6 +149,45 @@ static inline void sys_ring_buf_init(struct ring_buf *buf, u32_t size,
+static inline int sys_ring_buf_custom_space_get(u32_t size, u32_t head,
+						u32_t tail)
+{
+	if (sys_ring_buf_custom_is_empty(head, tail)) {
```
Why this special case? Seems not needed.

nordic-krch replied:
removed, in fact seems not needed.

nordic-krch commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +80,123 @@ int sys_ring_buf_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+		return -EINVAL;
+	}
+
+	buf->tail = (buf->tail + size) % buf->size;
```
well, yes but it's done one per chunk and not on every byte. Would you prefer something like:
`buf->tail = (buf->tail + size);`
`buf->tail = buf->tail > buf->size ? buf->tail - buf->size : buf->tail;`

nordic-krch replied:
fixed

pfalcon replied:
@nordic-krch :

Sorry, I guess I should have mentioned I'm packing for travel/will travel, so may be slow with responses.

> Would you prefer something like:
> buf->tail = buf->tail > buf->size ? buf->tail - buf->size : buf->tail;

Yes, that's what I hinted at. You see, it was easier with your previous version, which worked char by char ;-). Let me have a look at what you have now.

pfalcon replied:
> well, yes but it's done one per chunk and not on every byte.

As I mentioned, my usecase is console subsys, which works byte by byte, so would be affected. (Yes, it might be optimized, but that's a separate task, and a some (much of?) hardware still have byte by byte IRQ).

nordic-krch commented:
@pfalcon regarding API naming. As this ringbuf is inspired (compied) by ringbuffer we have in nrf5 sdk i followed naming convention we had there (alloc, put, get, free) but if something clearer comes out im open to change it. 

I see following options:
- alloc -> claim (because user claims the buffer for writing)
- put -> commit (because user commits actual amount of valid data)
- get -> fetch
- free -> release (because you cannot have free without alloc ;) )

nordic-krch commented on include/ring_buffer.h:
```diff
@@ -116,6 +149,45 @@ static inline void sys_ring_buf_init(struct ring_buf *buf, u32_t size,
+	}
+
+	if (tail < head) {
+		return head - tail - 1;
```
seems ok, if head is ahead of tail in the buffer then free space is between head and tail, size is not needed.

nordic-krch commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +80,123 @@ int sys_ring_buf_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+		return -EINVAL;
+	}
+
+	buf->head = (buf->head + size) % buf->size;
```
+1 for fixed

pfalcon commented:
> >    e.g. pass buffer to DMA
> That's quite a turn for this API! ;-)

So, thinking about it more, that's really a hasted change. The original patch already made enough change to API/code of the existing feature. And there're enough there to scrutinize and consider alone.

This new feature of "alloc'ing" really should have been a separate PR, not even just a separate commit. And yet it's now part of the same commit. So, rhetoric request - if it's not too late, please split "alloc" stuff into a separate commit with an outlook for separate PR.

pfalcon commented:
> regarding API naming.

Thanks for discussion on this. My immediate idea would be "why get and put use different terminology?" Why it can't be put_claim()/put_finish() and get_claim()/get_finish(). Specific verbs are worth consideration of course.

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -21,27 +22,66 @@ by its memory address.
+Byte mode
+=========
+
+Ring buffer instance works in byte mode if declared using
```
Similar change:

    A **byte mode** ring buffer instance is declared using

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -21,27 +22,66 @@ by its memory address.
+:cpp:func:`SYS_RING_BUF_BYTES_DECLARE_SIZE()` and accessed using:
+:cpp:func:`sys_ring_buf_bytes_alloc()`, :cpp:func:`sys_ring_buf_bytes_put()`,
+:cpp:func:`sys_ring_buf_bytes_get()`, :cpp:func:`sys_ring_buf_bytes_free()`,
+:cpp:func:`sys_ring_buf_bytes_cpy_put()` and
```
I'm not fond of abbreviations that remove only one letter (``cpy`` instead of ``copy``), but I'll leave that to the developer reviewers to discuss :)

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -21,27 +22,66 @@ by its memory address.
+
+Data can be copied into the ring buffer (see
+:cpp:func:`sys_ring_buf_bytes_cpy_put()`) or ring buffer memory can be used
+directly by the user. In the latter case, operation is splitted into three
```
change to (note the blank line needed before a list):

    In the latter case, the operation is split into three stages:

    1. allocating the buffer (:cpp:func:`sys_ring_buf_bytes_alloc()`) when
       user requests the destination location where data can be written.
    #. writing the data by the user (e.g. buffer written by DMA).
    #. indicating the amount of data written to the provided buffer
       (:cpp:func:`sys_ring_buf_bytes_put()`). The amount
       can be less than or equal to the allocated amount.

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -21,27 +22,66 @@ by its memory address.
+ (:cpp:func:`sys_ring_buf_bytes_put()`). It can be equal or less than allocated
+ amount.
+
+Data can be retrieved from a ring buffer through copying
```
And change this paragraph to:

    Data can be retrieved from a ring buffer through copying
    (see :cpp:func:`sys_ring_buf_bytes_cpy_get()`) or accessed
    directly by address. In the latter case, the operation is split
    into three stages:

    1. retrieving source location with valid data written to a ring buffer
       (see :cpp:func:`sys_ring_buf_bytes_get()`).
    #. processing data
    #. freeing processed data (see :cpp:func:`sys_ring_buf_bytes_free()`). 
       The amount freed can be less than or equal or to the retrieved amount.

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -74,7 +115,7 @@ Defining a Ring Buffer
 It must then be initialized by calling :cpp:func:`sys_ring_buf_init()`.
 
-The following code defines and initializes an empty ring buffer
+The following code defines and initializes an empty ring buffer for data items
```
change to ``The following code defines and initializes an empty **data item mode** ring buffer``

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -114,6 +155,15 @@ which can be accessed using less efficient modulo operations.
     #define MY_RING_BUF_WORDS 93
     SYS_RING_BUF_DECLARE_SIZE(my_ring_buf, MY_RING_BUF_WORDS);
 
+The following code defines a ring buffer with an arbitrary-sized data buffer,
```
How about instead:

    The following code defines an application-specific sized **byte mode** ring buffer
    enqueued and dequeued as raw bytes:

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -144,6 +194,42 @@ can be specified.
 	...
     }
 
+Bytes are copied to a ring buffer by calling
```
For clarity change this to:

    Bytes are copied to a **byte mode** ring buffer by calling

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -144,6 +194,42 @@ can be specified.
     }
 
+Bytes are copied to a ring buffer by calling
+:cpp:func:`sys_ring_buf_bytes_cpy_put()`.
```
After the period, add:

    For example:

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -144,6 +194,42 @@ can be specified.
+	...
+    }
+
+Data can be added to a ring buffer by direct operation on a ring buffer memory.
```
How about:

    Data can be added to a **byte mode** ring buffer by directly access the
    ring buffer's memory.  For example:

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -170,6 +256,47 @@ A data item is removed from a ring buffer by calling
         ...
     }
 
+Bytes data is copied out from a ring buffer by calling
```
Change to:

    Data bytes are copied out from a **byte mode** ring buffer by calling

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -170,6 +256,47 @@ A data item is removed from a ring buffer by calling
     }
 
+Bytes data is copied out from a ring buffer by calling
+:cpp:func:`sys_ring_buf_bytes_cpy_get()`.
```
After the period, add:

    For example:

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -170,6 +256,47 @@ A data item is removed from a ring buffer by calling
+        ...
+    }
+
+Data can be retrieved from a ring buffer by direct operations on a ring buffer
```
How about:

    Data can be retrieved from a **byte mode** ring buffer by direct
    operations on the ring buffer's memory.  For example:

nordic-krch commented:
@pfalcon regarding:
> So, thinking about it more, that's really a hasted change. 

cpy_put, cpy_get are just wrappers around 2step approach so i'm a bit reluctant to split it into two PR's. Especially, that im doing it to add more robust transport in new shell. Initially, i thought that i'll start will straightforward but not optimal copy-ring-buffer but with uart irq api it is actually much more convenient to have zero-copy-ring-buffer. We had those zero-copy-ring-buffers in nrf sdk and they were very convenient for streams use cases like uart.

> put_claim()/put_finish() and get_claim()/get_finish()

I like that with one 'but': _sys_ring_buf_bytes_put_claim_ is too long and too confusing. So I have question to @nashif : do we accept api changes which are basic renames? If yes then I would have following proposal:
- remove sys_ prefix since it seems useless
- sys_ring_buf_put/get -> ring_buf_item_put/get (byte ring buffer seems more common than packet ring buffer)
- sys_ring_buf_bytes_ prefix -> ring_buf_

Note that currently ring buffer is barely used (ipm_console_receiver.c and event_logger.c).

pfalcon commented:
> Especially, that im doing it to add more robust transport in new shell. ... We had those zero-copy-ring-buffers in nrf sdk and they were very convenient for streams use cases like uart.

Thanks, it's always helpful to know the usecases behind the features proposed.

Let me also step back and give some explanations which seem implied to me, but may be not clear overall. So, I hope you know that I'm a guy who wrote "yet another ring buffer": https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/console/getchar.c#L18 . So, I'm really excited to see that you didn't go that way, and decided to resolve it for everyone. And of course, I'd be looking to replace adhoc ringbuf in subsys/console, and that's my usecase, and I review this PR with that outlook, to ensure there're would be no "regressions". E.g., my impl doesn't use %, that's why my repeating comments about it.

But if we really want to improve state of Zephyr by getting rid of adhoc implementations in favor of a generic/general one, we should be sure we're doing right thing on a wider scale. And here a point which I noticed recently.

As a prelude to it, I have to admit that I can't say that I'm well versed with Zephyr's "IPC" primitives. I know there're bunch of them, but I mostly use what other people use in the code I work with, and I definitely don't know detailed differences among them (beyond the obvious ones). So, I recently came against https://docs.zephyrproject.org/latest/kernel/data_passing/pipes.html . It works in terms of bytes and:

> The pipe can be configured with a ring buffer which holds data that has been sent but not yet received; 

So, let's pause for a bit with the rest of review and answer the question: Do we do right thing by patching ring_buffer object? Why can't we use k_pipe instead?

nordic-krch commented:
@pfalcon I've seen pipes and also thought that it's a ringbuffer in a way but also much more:
 > A pipe is a kernel object that allows a thread to send a byte stream to another thread. 
And when i look into the implementation, it is heavy. Too heavy for my use case since i want shell to be able to process a lot of data (logs). Maybe pipe should be built on top of the ring buffer (@nashif does it have to be part of a kernel then?)

pfalcon commented:
> I've seen pipes

Great!

> the implementation, it is heavy. Too heavy for my use case

Ack. Moreover, from docs:

> The kernel does NOT allow for an ISR to send or receive data to/from a pipe even if it does not attempt to wait for space/data.

And we definitely need ISR access to ring_buf. (More then that, you want to do DMA straight to them!)


So, we're on the same line that we don't reinventing the wheel. (But may need to adjust k_pipe later, ack. Can think about it later indeed.)

pfalcon commented:
But I took a chance to submit https://github.com/zephyrproject-rtos/zephyr/issues/9936 on that (that also can shed light why most primitives have k_ prefix, while ring_buf suddenly sys_).

nordic-krch committed:
lib: ring_buffer: Rename sys_ring_buf_ to ring_buf_item_

Deprecate API prefixed with sys_ring_buf_ and rename it
to ring_buf_item_ since this API is not a typical ring buffer
but ring buffer of data items (metadata + 32bit words).

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@pfalcon @nashif haven't heard no about current ring_buf API so I went ahead and deprecated current ring buffer api (commit https://github.com/zephyrproject-rtos/zephyr/pull/9853/commits/595e91be1e3b5723a066aab7dca230b416b55e3a). I've renamed it (keeping old ones with __deprecated prefix) by removing _sys__ prefix (which was not aligned with file names and was simply redundant) and adding __item_ infix to indicate that it's not standard ring buffer API.
Rename snippets:
- sys_ring_buf_put -> ring_buf_item_put
- sys_ring_buf_init -> ring_buf_init

That allowed to have functions _ring_buf_put_, _ring_buf_get_ which do what is expected from them (copy in/out bytes to a ring buffer).

pfalcon commented on include/ring_buffer.h:
```diff
@@ -242,6 +287,142 @@ __deprecated static inline int sys_ring_buf_get(struct ring_buf *buf,
+u32_t ring_buf_get_claim(struct ring_buf *buf, u8_t **data, u32_t size);
+
+/**
+ * @brief Indicate number of bytes that can be freed.
```
IMHO this description is a bit weak and unclear.

Compare it with description for put_finish: 

> Indicate number of bytes written to allocated buffers.


Would be nice if it also talk about bytes read, etc.

pfalcon commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +80,136 @@ int ring_buf_item_get(struct ring_buf *buf, u16_t *type, u8_t *value,
 	return 0;
 }
+
+/** @brief Wraps if exceeds the limit.
```
Another pretty quick a description, even for brief. Maybe at least "Wrap index if it exceeded the limit"?

pfalcon commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +80,136 @@ int ring_buf_item_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+	space = z_ring_buf_custom_space_get(buf->size, buf->head,
+					    buf->misc.byte_mode.tmp_tail);
+
+	/* Limit requested size to available size. */
```
Aha, that's cool that you have such a check.

pfalcon commented on lib/ring_buffer/ring_buffer.c:
```diff
@@ -79,16 +80,136 @@ int ring_buf_item_get(struct ring_buf *buf, u16_t *type, u8_t *value,
+					    buf->misc.byte_mode.tmp_tail);
+
+	/* Limit requested size to available size. */
+	size = size > space ? space : size;
```
Btw, there's a min() func ;-).

pfalcon commented:
> @pfalcon @nashif haven't heard no about current ring_buf API so I went ahead and deprecated current ring buffer api

All I can say is that if it is to be done, it should be done just like you did - with providing aliases for older names, which can be marked as deprecated.

I looked thru the full of the new patch, and would say that new names look pretty consistent and good. @nordic-krch, I hope you yourself is satisfied with them, because my idea was to just propose some alternatives, so we were sure that we considered various alternatives. But again, I really like these put/get claim/finish pairs.

There're some suggestions still to use min() instead ?: and possibly improve some docstrings, but to signify progress, let me dismiss my -1. Thanks for all the refactors!

nordic-krch commented:
@pfalcon @dbkinder thank you for review, comments addressed.
Ping to other reviewers who may want to take a look.
@nashif  @carlescufi could comment (hopefully approve) deprecation of sys_ring_buf_...?

pfalcon reviewed (approved):
@nordic-krch: Thanks for all the changes. I assume other folks don't have issues with this PR, so let me approve it.

@dbkinder : Can you please recheck the docs?

carlescufi commented on include/ring_buffer.h:
```diff
@@ -78,42 +96,92 @@ struct ring_buf {
- * used for ring buffers not defined using SYS_RING_BUF_DECLARE_POW2 or
- * SYS_RING_BUF_DECLARE_SIZE.
+ * used for ring buffers not defined using SYS_RING_BUF_DECLARE_POW2,
+ * SYS_RING_BUF_DECLARE_SIZE or SYS_RING_BUF_RAW_DECLARE_SIZE.
```
`SYS_RING_BUF_RAW_DECLARE_SIZE` doesn't exist?

nordic-krch replied:
fixed

carlescufi commented on tests/lib/ringbuffer/src/main.c:
```diff
@@ -227,6 +230,167 @@ void test_ringbuffer_size_put_get_thread_isr(void)
+	}
+}
+
+void test_raw_put_free(void)
```
shouldn't this be `test_byte_put_free` ?

nordic-krch replied:
yep, renamed.

pfalcon commented:
@nordic-krch: ping

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -138,17 +193,54 @@ can be specified.
 	...
     }
 
+ytes are copied to a **byte mode** ring buffer by calling
```
s/ytes/Bytes/

nordic-krch replied:
done

dbkinder commented on doc/kernel/other/ring_buffers.rst:
```diff
@@ -138,17 +193,54 @@ can be specified.
+	...
+    }
+
+Data can be added to a **byte mode** ring buffer by directly access the
```
s/access/accessing/

nordic-krch replied:
done

dbkinder reviewed (dismissed):
a couple of doc typos...

nordic-krch commented:
@dbkinder @carlescufi i've fixed your findings.

dbkinder reviewed (approved):
LGTM, thanks.
